### PR TITLE
feat: add support to install different architectures

### LINKED ISSUE #5965: Cross platform / architecture install option
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

npm install offers option to select a different platform/architecture. This is useful for example in case packages have a post-install script that fetches arch-specific binaries (such as sharp).

See https://sharp.pixelplumbing.com/install#cross-platform for an example.

**Describe the solution you'd like**

`pnpm install --platform=linux --architecture=x64`

**Describe the drawbacks of your solution**

I don't think there are any drawbacks. NPM has implemented this, but may have taken some shortcuts that I am not aware of and pnpm project isn't willing to take.

**Describe alternatives you've considered**

Spinning up a machine (or docker container) with that architecture and run pnpm install in that. But that's a rather heavyweight solution to this problem.

### PR DESCRIPTION
**Pull request by nachoaldamav**
This feature allows pnpm to force the install of dependencies for specific systems.

For example, with this `package.json`:
```json
{
  "pnpm": {
    "supportedArchitectures": {
      "os": ["darwin", "win32"],
      "cpu": ["x64", "arm64"]
    }
  }
}
```

pnpm will install the following dependencies for esbuild:
```bash
  - @esbuild+darwin-arm64@0.18.20
  - @esbuild+darwin-x64@0.18.20
  - @esbuild+win32-arm64@0.18.20
  - @esbuild+win32-x64@0.18.20
```

TODO:
- [x] add tests for supportedArchitectures and hoisted node_modules
- [x] add test for the case when the optional deps are in subdeps, architectures are changed and a new dependency is installed.
- [x] verify that the old optional deps are removed when the value of supportedArchitectures is changed.

close #5965

### CODE REVIEW

File: cli/cli-utils/src/readProjectManifest.ts [hunk 1]
```diff
@@ -23,9 +24,13 @@ export async function readProjectManifestOnly (
   opts: {
     engineStrict?: boolean
     nodeVersion?: string
-  } = {}
+    supportedArchitectures?: SupportedArchitectures
+  } = {
+
+  }
```
Comment (zkochan) [inline cli/cli-utils/src/readProjectManifest.ts]: avoid change to formatting unrelated to the changes
```suggestion
  } = {}
```

File: cli/cli-utils/src/readProjectManifest.ts [hunk 2]
```diff
@@ -23,9 +24,13 @@ export async function readProjectManifestOnly (
...
 ): Promise<ProjectManifest> {
   const manifest = await utils.readProjectManifestOnly(projectDir)
+
```
Comment (zkochan) [inline cli/cli-utils/src/readProjectManifest.ts]: avoid change to formatting unrelated to the changes

File: config/package-is-installable/src/checkPlatform.ts [hunk 3]
```diff
@@ -45,25 +53,44 @@ export interface Platform {
...
       }
-      ++blc
-    } else {
-      match = match || tmp === value
     }
   }
   return match || blc === list.length
 }
+
+function dedupeArchs (current: string, supported: string[]) {
```
Comment (zkochan) [inline config/package-is-installable/src/checkPlatform.ts]: This name of the function isn't good as it is used not only for arch. Maybe
```suggestion
function dedupeCurrent (current: string, supported: string[]) {
```

File: config/package-is-installable/src/checkPlatform.ts [hunk 4]
```diff
@@ -45,25 +53,44 @@ export interface Platform {
...
+  const result = supported.filter((arch) => arch !== 'current')
+
+  if (supported.includes('current')) {
+    result.push(current)
+  }
+
+  return result
```
Comment (zkochan) [inline config/package-is-installable/src/checkPlatform.ts]: I guess a more effective way would be to do something like this
```suggestion
  return supportedList.map((supported) => supported === 'current' ? current : supported)
```

File: exec/plugin-commands-script-runners/test/index.ts [hunk 5]
```diff
@@ -29,6 +29,7 @@ test('pnpm run: returns correct exit code', async () => {
     extraBinPaths: [],
     extraEnv: {},
     rawConfig: {},
+
```
Comment (zkochan) [inline exec/plugin-commands-script-runners/test/index.ts]: revert the new lines in the file

File: config/package-is-installable/test/checkPlatform.ts
Comment (zkochan) [inline config/package-is-installable/test/checkPlatform.ts:1]: add some tests that also check not by `current`

**Review by zkochan — CHANGES_REQUESTED** [review]

**Comment by zkochan** [comment]
We should also ensure that if someone modifies the `supportedArchitectures` field, pnpm will install the new included artifacts and symlink them to the dependent packages.

**Comment by nachoaldamav** [comment]
> We should also ensure that if someone modifies the `supportedArchitectures` field, pnpm will install the new included artifacts and symlink them to the dependent packages.

I just tested something, install everything, change the field and install again (keep `pnpm-lock.yaml` and `node_modules`), the new ones will be added but not symlinked, it won't remove anything that was removed from the `package.json`.

**Review by zkochan — APPROVED** [review]

**Comment by Doni108doni108** [comment]
Leave a comment

**Comment by segevfiner** [comment]
@nachoaldamav Why is this in the `package.json` though? Doesn't this make more sense as configuration that can also be set via CLI and config files?

**Pull request merged**